aptos-move/aptos-vm/src/transaction_arg_validation.rs

```
@@ -83,22 +101,133 @@ pub(crate) fn validate_combine_signer_and_txn_args<S: MoveResolverExt>(
+fn validate_args<S: MoveResolverExt>(
+    session: &SessionExt<S>,
+    idxs: &[usize],
+    args: &[Vec<u8>],
+    func: &LoadedFunctionInstantiation,
+) -> Result<(), VMStatus> {
+    let mut current_size_for_strings = 0;
+    for (idx, (ty, arg)) in func.parameters.iter().zip(args.iter()).enumerate() {
+        if !idxs.contains(&idx) {
+            continue;
+        }
+        let mut cursor = Cursor::new(&arg[..]);
+        validate_arg(session, ty, &mut cursor, &mut current_size_for_strings)?;
+    }
+    Ok(())
+}
+
+// Validate a single arg. A Cursor is used to walk the serialized arg manually and correctly.
+// Only Strings and nested vector of them are valiadated.
+fn validate_arg<S: MoveResolverExt>(
+    session: &SessionExt<S>,
+    ty: &Type,
+    cursor: &mut Cursor<&[u8]>,
+    current_size_for_strings: &mut usize,
+) -> Result<(), VMStatus> {
+    use move_deps::move_vm_types::loaded_data::runtime_types::Type::*;
+    Ok(match ty {
+        Vector(inner) => {
+            // get the vector length and iterate over each element
+            let mut len = get_len(cursor)?;
+            *current_size_for_strings += len;
+            if *current_size_for_strings > MAX_STRING_ARGS_SIZE {
+                return Err(VMStatus::Error(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT));
+            }
+            while len > 0 {
+                validate_arg(session, inner, cursor, current_size_for_strings)?;
+                len -= 1;
+            }
+        }
+        // only strings are validated, and given we are here only if one was present
+        // (`is_valid_txn_arg`), this match arm must be for a string
+        Struct(idx) | StructInstantiation(idx, _) => {
+            // load the struct name, we use `expect()` because that check was already
+            // performed in `is_valid_txn_arg`
+            let st = session
+                .get_struct_type(*idx)
+                .expect("unreachable, type must exist");
```

this and below, should we return a failure instead?
this is the impossible condition I mentioned i the main comment, but....